Implement nrf_cloud_fota (fota_v2)

jayteemo opened the pull request:
Replace aws_fota (AWS Jobs based) in the nrf_cloud library with nrf_cloud_fota, which is custom for nRF Cloud.
A few asset_tracker changes required for using nrf_cloud_fota.

CIA-64

sigvartmh commented on applications/asset_tracker/prj.conf:
```diff
@@ -90,24 +86,22 @@ CONFIG_HEAP_MEM_POOL_SIZE=16384
 CONFIG_MCUBOOT_IMG_MANAGER=y
-
-# Flash
-CONFIG_FLASH=y
```
Nit: Align the changes if possible so it's a bit easier to see. Currently, it looks like you do changes to the flash configuration.

sigvartmh replied:
It's a nit so feel free to ignore it :)

sigvartmh commented on applications/asset_tracker/src/main.c:
```diff
@@ -1377,7 +1377,7 @@ void cloud_event_handler(const struct cloud_backend *const backend,
 		ui_led_set_pattern(UI_CLOUD_CONNECTED);
 
-#if defined(CONFIG_BOOTLOADER_MCUBOOT)
+#if defined(CONFIG_BOOTLOADER_MCUBOOT) && !IS_ENABLED(CONFIG_NRF_CLOUD_FOTA)
```
use `defined` `IS_ENABLED` macro AFAIK is reserved for using it in a C `if` and not preprocessor.

The reason for using IS_ENABLED in a regular `if` is that you can catch function signature changes etc. The code path will be compiled but not linked.

sigvartmh commented on applications/asset_tracker/src/main.c:
```diff
@@ -1750,7 +1750,7 @@ static void ui_evt_handler(struct ui_evt evt)
 void handle_bsdlib_init_ret(void)
 {
-#if defined(CONFIG_BSD_LIBRARY)
+#if defined(CONFIG_BSD_LIBRARY) && !IS_ENABLED(CONFIG_NRF_CLOUD_FOTA)
```
This seems to change something more fundamental with the application but I understand why it is needed.

sigvartmh commented on subsys/net/lib/nrf_cloud/Kconfig:
```diff
@@ -103,13 +103,32 @@ config NRF_CLOUD_MQTT_PAYLOAD_BUFFER_LEN
 config NRF_CLOUD_FOTA_PROGRESS_PCT_INCREMENT
 	int "Percentage increment at which FOTA download progress is reported"
-	depends on FOTA_DOWNLOAD_PROGRESS_EVT
+	depends on NRF_CLOUD_FOTA
```
```suggestion
	depends on NRF_CLOUD_FOTA || FOTA_DOWNLOAD_PROGRESS_EVT

sigvartmh replied:
If the pure AWS Jobs solution is completely removed, then it won't be needed. If it's not removed then this is needed for having the possibility of enabling it.

sigvartmh commented on subsys/net/lib/nrf_cloud/include/nrf_cloud_fota.h:
```diff
@@ -0,0 +1,151 @@
+	NRF_FOTA_TYPE__FIRST = 0,
+
+	/** Application update. */
+	NRF_FOTA_APPLICATION = NRF_FOTA_TYPE__FIRST,
```
Why is it assigned like this?

sigvartmh commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,928 @@
+	 * [“jobExecutionId”,firmwareType,fileSize,”host”,”path”]
+	 *
+	 * Example:
+	 * [“abcd1234”,0,1234,”nrfcloud.com”,"v1/firmwares/appfw.bin"]
```
So this looks like a JSON array to me. If cJSON already is used by nRFCloud may I suggest maybe trying to rewrite the parsing logic using cJSON?

Could be that it's not possible it's just a suggestion.

```
int status;
cJSON_ArrayForEach(token, tokens) {
        if (!cJSON_IsNumber(token)) {
            status = 0;
            goto handle_error;
        }
    }

jayteemo replied:
probably possible, i can try it out to see how it looks.  i did not consider pulling in a library since the parsing was simple.

sigvartmh replied:
Yeah, I think cJSON is already in the nRF Cloud library for doing some JSON parsing. So instead of having custom parsing logic (even if simple), it would be nice to abstract it away if possible.

sigvartmh commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,928 @@
+#define TOPIC_FOTA_REQ    "/jobs/req"
+#define TOPIC_FOTA_UPDATE "/jobs/update"
+
+#define JOB_UPDATE_MSG_TEMPLATE "[\"%s\",%d,%s]"
```
Since we are writing our own custom thing. Why are we using JS arrays instead of CBOR?

jayteemo replied:
CBOR was not considered.  it looks like it would only save us ~10 bytes per message?  if there are compelling reasons to implement CBOR for this then it should be done.

i didn't think to use cJSON since the payloads are very simple.  i can try it out if you think it provides a benefit?

sigvartmh replied:
I don't know the details of the implementation. If the cbor library is not already included in the application, I would say no since it would add to 1828 bytes, pr. update, which probably will waste the 10 bytes pr message save. An additional advantage of cbor would be parsing speed as it's lighter to parse.

If it's already included in the application, I would say go for it. Anything which can make the transport overhead lower is a great improvement.

If you can write a few generic functions with cJSON, maybe it would save a few bytes of flash.

sigvartmh replied:
Remember to build with optimize for size when checking this.

sigvartmh commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,928 @@
+	enum fota_validate_status ret = NRF_FOTA_VALIDATE_UNKNOWN;
+
+#if defined(CONFIG_BSD_LIBRARY)
+	int modem_dfu_res = bsdlib_get_init_ret();
```
Not the correct place for the change  just FYI:
You can now deinitialize and initialize the modem without having to power-cycle the chip to perform a modem update.

I think the sequence is
```
bsd_shutdown();
int modem_dfu_res;
check_result:
modem_dfu_res = bsd_init();
/* Handle return values relating to modem firmware update */
	switch (modem_dfu_res) {
	case MODEM_DFU_RESULT_OK:
		LOG_DBG("Modem FOTA OK");
		ret = NRF_FOTA_VALIDATE_PASS;
                 bsd_shutdown();
                 goto check_result;
	case MODEM_DFU_RESULT_UUID_ERROR:
	case MODEM_DFU_RESULT_AUTH_ERROR:
	case MODEM_DFU_RESULT_HARDWARE_ERROR:
	case MODEM_DFU_RESULT_INTERNAL_ERROR:
		LOG_ERR("Modem FOTA error: %d", modem_dfu_res);
		ret = NRF_FOTA_VALIDATE_FAIL;
		break;
	default:
		LOG_DBG("Modem FOTA result unknown: %d", modem_dfu_res);
		break;
	}

jayteemo replied:
ah nice, good to know.

sigvartmh replied:
I think it's bsdlib specific

jayteemo replied:
i need to rebase and retest, initially this did not work as expected.

sigvartmh commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:
```diff
@@ -664,24 +523,68 @@ static int nct_settings_init(void)
 	int ret = 0;
 
-#if !IS_ENABLED(CONFIG_MQTT_CLEAN_SESSION)
+#if !IS_ENABLED(CONFIG_MQTT_CLEAN_SESSION) || IS_ENABLED(CONFIG_NRF_CLOUD_FOTA)
```
nit: IS_ENABLED
use define instead.

sigvartmh replied:
If you want to use IS_ENABLED it should be done with 
```
if(!IS_ENABLED(CONFIG_MQTT_CLEAN_SESSION) || IS_ENABLED(CONFIG_NRF_CLOUD_FOTA)
```

So that the following code is checked.

sigvartmh commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:
```diff
@@ -664,24 +523,68 @@ static int nct_settings_init(void)
 	return ret;
 }
 
+#if IS_ENABLED(CONFIG_NRF_CLOUD_FOTA)
```
nit: use `defined`

sigvartmh commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:
```diff
@@ -715,14 +618,6 @@ int nct_mqtt_connect(void)
 		nct.client.transport.type = MQTT_TRANSPORT_NON_SECURE;
 #endif
-#if defined(CONFIG_AWS_FOTA)
-		err = aws_fota_init(&nct.client, aws_fota_cb_handler);
```
Is support for FOTA_V1 removed?

sigvartmh replied:
Can this number be found on the confluence page of FOTA v2? if not add it for reference :)

jayteemo replied:
added.
`19 users, averaging 4.5 jobs in 2020`

jtguggedal commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,928 @@
+	case MQTT_EVT_PUBREC:
+	case MQTT_EVT_PUBREL:
+	case MQTT_EVT_PUBCOMP:
+	case MQTT_EVT_PINGRESP:
```
Perhaps add a default case too, in the unlikely event that a new MQTT event is added upstream

jtguggedal commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,928 @@
+	__ASSERT_NO_MSG(job != NULL);
+
+	int ret;
+	int sec_tag = -1;
```
Should customer-provided sec tag be supported? If they use HTTPS and hosting their images at a server that doesn't use AWS it's needed

jayteemo replied:
for nrf cloud fota, all files will be served by nrf cloud (aws)

jtguggedal commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,928 @@
+
+#if IS_ENABLED(CONFIG_NRF_CLOUD_FOTA_HTTPS_DOWNLOADS)
+	sec_tag = CONFIG_NRF_CLOUD_SEC_TAG;
+	fragment_size = 1024;
```
I think this can be increased, depending on the typical HTTP header size from AWS. Perhaps it should even be provided in the job description somehow to allow customers to fine-tune the fragment size to decrease downloaded bytes and increase speed

jayteemo replied:
yes i agree this can be increased (or made into a config option).  once we get fota v2 pushed to dev or beta, i will test with a larger size.

jtguggedal reviewed (commented):
Did a first pass. Looks good!
Added a few comments and suggestions to consider

jayteemo commented on subsys/net/lib/nrf_cloud/include/nrf_cloud_fota.h:
```diff
@@ -0,0 +1,151 @@
+ *        This should be used during initialization.
+ * @param client      Pointer to the MQTT client instance.
+ * @param client_id   Client id of this device.
+ * @param endpoint    User and device specific MQTT endpoint.
```
this is the base endpoint upon which the fota topics will be constructed.

jayteemo commented on subsys/net/lib/nrf_cloud/include/nrf_cloud_fota.h:
```diff
@@ -0,0 +1,151 @@
+ * @retval 0 If successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nrf_cloud_fota_endpoint_set(struct mqtt_client *const client,
```
as noted, `nrf_cloud_fota_endpoint_set_and_report()` should be used during init.
`nrf_cloud_fota_endpoint_set` probably isn't necessary; i can't think of an occasion when the endpoint would change after init.

jayteemo commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,928 @@
+		return -EINVAL;
+	}
+
+	event_cb = cb;
```
yes, not sure why anyone would need to do that, but they could 🤷‍♂️

jayteemo commented:
i updated with a couple TODO items, but if you want to try it out on dev.nrfcloud.com you can.
there are two modem fw updates 1.2.1->1.2.1_FOTA-TEST and 1.2.2->1.2.2_FOTA-TEST.
you should be able to create your own app update.

jayteemo commented:
removed draft status.  nrf cloud will be pushing fota_v2 to prod at the end of next week.
still waiting for https://github.com/nrfconnect/sdk-nrf/pull/3436 to be merged... and a one line change will be required to `fota_download_start` to set the new `set_tls_hostname` flag.
is there any reason to *not* set the TLS hostname in `fota_download_start`?  if there is, then it would need to be added as a parameter.

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1217 @@
+{
+	cJSON *item = cJSON_CreateString(string);
+
+	if (item && array) {
```
If `array` is NULL we have a memory leak (no one frees `item`). 

I know it's a static function and it's likely not expected to provide a NULL `array` pointer, but it would be more readable to verify it serparately (with an assert for instance).

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1217 @@
+{
+	cJSON *item = cJSON_CreateNumber(number);
+
+	if (item && array) {
```
Same story here.

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1217 @@
+	struct nrf_cloud_fota_ble_job ble_job;
+
+	if (strstr(sub_topics[SUB_TOPIC_IDX_BLE_RCV].topic.utf8,
+			p->message.topic.topic.utf8) != NULL) {
```
Please align with the opening `strstr` bracket.

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1217 @@
+#endif
+
+	if (strstr(sub_topics[SUB_TOPIC_IDX_RCV].topic.utf8,
+	    p->message.topic.topic.utf8) == NULL && !ble_id) {
```
Please align with the opening` strstr` bracket.

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1217 @@
+					   p->message.payload.len);
+	if (ret) {
+		LOG_ERR("Error reading MQTT payload: %d", ret);
+		goto send_ack;
```
`payload` leaks here.

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1217 @@
+
+	ret = parse_job_info(job_info, ble_id, payload, &payload_array);
+	if (ret) {
+		goto send_ack;
```
`payload` leaks here as well.

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:
```diff
@@ -664,24 +523,68 @@ static int nct_settings_init(void)
+				      * const evt)
+{
+	switch (evt->id) {
+	case NRF_CLOUD_FOTA_EVT_START:
+	{
```
```suggestion
	case NRF_CLOUD_FOTA_EVT_START: {
```
Here and all below.

rlubos commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1217 @@
+	/* Report status of saved job now that the endpoint is available */
+	if (saved_job.type != NRF_CLOUD_FOTA_TYPE__INVALID) {
+
+		struct nrf_cloud_fota_job job = {
+			.info = { .type = saved_job.type,
+				  .id = saved_job.id
+				}
+			};
+
```
```suggestion
		struct nrf_cloud_fota_job job = {
			.info = { 
				.type = saved_job.type,
				.id = saved_job.id
			}
		};

```

plskeggs commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1226 @@
+			return -EFAULT;
+		}
+		LOG_DBG("Subscribing to topic: %s",
+			sub_list.list[i].topic.utf8);
```
missing a call to log_strdup() here

jayteemo commented:
added change required for SNI in `fota_download`.  let me know if you would prefer that to be a `fota_download` kconfig or function parameter addition to `fota_download_start`
updated doc.
removed DNM tag.

jayteemo commented:
fota_v2 is now live on [nrfcloud.com](nrfcloud.com) 🎈

jtguggedal commented on applications/asset_tracker/src/main.c:
```diff
@@ -1790,8 +1790,8 @@ static void ui_evt_handler(struct ui_evt evt)
-#if defined(CONFIG_NRF_MODEM_LIB)
-	int ret = nrf_modem_lib_get_init_ret();
+#if defined(CONFIG_NRF_MODEM_LIB) && !defined(CONFIG_NRF_CLOUD_FOTA)
+	int ret = bsdlib_get_init_ret();
```
This has now changed to `nrf_modem_lib_get_init_ret()`

rlubos replied:
Too bad it wasn't caught by the CI... Is it ok to build asset_tracker w/o `NRF_CLOUD_FOTA` enabled? If so, perhaps we should add such a variant to the `sample.yaml`?

jayteemo replied:
yes it is ok to build w/o fota.  
by default fota is enabled when nrf_cloud is enabled.

jtguggedal commented on subsys/net/lib/nrf_cloud/Kconfig:
```diff
@@ -103,13 +103,35 @@ config NRF_CLOUD_MQTT_PAYLOAD_BUFFER_LEN
+config NRF_CLOUD_FOTA_DOWNLOAD_FRAGMENT_SIZE
+	int "Fragment size for FOTA downloads."
+	range 128 2000
+	default 1024
```
Would it be safe to bump this up a bit to reduce the bytes sent back and forth? Depending on the HTTP header, perhaps something in the range 1700-1800 could work, but the risk is that the header grows over time, making the device un-FOTAable. In Azure FOTA this is worked around by adding the option to ship over the fragment size in the FOTA job document

jayteemo replied:
i'll test with some larger values.



> In Azure FOTA this is worked around by adding the option to ship over the fragment size in the FOTA job document

can you share the details of this?  i can see about getting that added to nrfcloud fota

jayteemo replied:
headers from modem update files are ~250 bytes... application update files are ~520 bytes 😅

jayteemo replied:
@jtguggedal  i've updated to 1700... which gives about 100 bytes of cushion.

however, something i don't quite understand:
if i use the [download](https://github.com/nrfconnect/sdk-nrf/tree/master/samples/nrf9160/download) sample and configure it for https and 2048 fragment size, it downloads just fine... and the http header size is ~730 bytes.

jtguggedal replied:
> > In Azure FOTA this is worked around by adding the option to ship over the fragment size in the FOTA job document

> can you share the details of this? i can see about getting that added to nrfcloud fota

For Azure, we use the JSON schema that can be seen here: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/net/azure_fota.html
The fragment size is then specified in the JSON document that described the FOTA sort of similar to how AWS Jobs work, just way simpler.

> @jtguggedal i've updated to 1700... which gives about 100 bytes of cushion.
> however, something i don't quite understand:
> if i use the download sample and configure it for https and 2048 fragment size, it downloads just fine... and the http header size is ~730 bytes.

Is this with TLS? I would expect it to work on non-TLS connection as that has 4k buffers in the modem, and fail with TLS with such a large HTTP header.

jayteemo replied:
@jtguggedal yes, TLS is enabled for the sample, i added a print for the http header size in http_parse in download_client:

```
*** Booting Zephyr OS build v2.4.0-ncs1-1715-g3366927a5498  ***
Download client sample started
Waiting for network.. OK
Downloading https://file-examples-com.github.io/uploads/2017/10/file_example_JPG_100kB.jpg
HTTP header len: 724
[   2% ] |=                                                 | (2048/102117 bytes) - HTTP header len: 726
[   4% ] |==                                                | (4096/102117 bytes) - HTTP header len: 726
[   6% ] |===                                               | (6144/102117 bytes) - HTTP header len: 726
...
```
I suppose inspecting this in wireshark would give me more insight.

jtguggedal replied:
That's very confusing... Agree that Wireshark is probably needed to understand what's going on.

jtguggedal commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1251 @@
+#include <nrf_modem.h>
+#endif
+
+LOG_MODULE_REGISTER(nrf_cloud_fota, CONFIG_NRF_CLOUD_LOG_LEVEL);
```
Perhaps let the FOTA have its own log level

jtguggedal commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:
```diff
@@ -0,0 +1,1251 @@
+		if (saved_job.type == NRF_CLOUD_FOTA_MODEM) {
+			/* Reboot is required if type is still set to MODEM */
+			LOG_INF("Rebooting to complete modem FOTA");
+			sys_reboot(SYS_REBOOT_COLD);
```
Is it possible to propagate an event here to the app instead of rebooting in the library? Then the application can have the choice to reboot or just restart the modem and not the application

jayteemo replied:
yes, i will create a JIRA task to improve this.  
i still wasn't able to get the modem-restart-without-reboot working, so i'll try some more for that too.

for modem fota there are 2 reboots that occur.  the initial reboot, which allows for the modem to install.
then, this, the reboot after the successful installation.

jtguggedal replied:
Oh okay, I thought it was just the one reboot. I haven't tested the "warm reboot" of the modem myself, but sounds like a useful feature in these cases. But we might run into issues with stuff that has initialized and set up sockets during boot, that will probably fail if modem is rebooted but their init functions are not run again. For instance common AT interface and GPS driver.

jayteemo replied:
CIA-175 for reference

jtguggedal reviewed (approved):
Looks good! Left a few comments to consider, but nothing blocking

jayteemo commented:
@rlubos i believe i have addressed the review comments.  it would be good to get this merged since it is now live on nrfcloud.
